src/sqlancer/ASTBasedReducer.java

```diff
@@ -0,0 +1,470 @@
+                "OrderByElements", "Joins" };
+        private final String[] pullUpList = { "Where", "Having", "FromItem", "SelectItems", "GroupBy", "Joins" };
+        private final String[] descendList = { "Where", "Having", "FromItem", "SelectItems", "GroupBy" };
+
+        private String getterName(String astNodeName) {
+            return "get" + astNodeName;
+        }
+
+        private String setterName(String astNodeName) {
+            if (astNodeName.equals("GroupBy")) {
+                return "set" + astNodeName + "Element";
+            } else {
+                return "set" + astNodeName;
+            }
+        }
+
+        @Override
+        public void visit(PlainSelect plainSelect) {
+            // Remove Section
+            for (String astNodeName : removeList) {
+                try {
+                    Method nodeGetter = plainSelect.getClass().getMethod(getterName(astNodeName));
+                    Object astNode = nodeGetter.invoke(plainSelect);
+                    if (astNode == null) {
+                        continue;
+                    }
+                    Method nodeSetter = plainSelect.getClass().getMethod(setterName(astNodeName),
+                            nodeGetter.getReturnType());
+                    nodeSetter.invoke(plainSelect, new Object[] { null });
+                    updateStatements();
+                    if (!bugStillTriggers()) {
+                        nodeSetter.invoke(plainSelect, astNode);
+                        updateStatements();
+                    }
+                } catch (Exception ignoredException) {
+                }
+            }
+
+            // Pull Up Section
+            for (String astNodeName : pullUpList) {
+                try {
+                    Method nodeGetter = plainSelect.getClass().getMethod(getterName(astNodeName));
+                    Object astNode = nodeGetter.invoke(plainSelect);
+                    if (astNode == null) {
+                        continue;
+                    }
+                    Method nodeSetter = plainSelect.getClass().getMethod(setterName(astNodeName),
+                            nodeGetter.getReturnType());
+
+                    if (astNode instanceof Expression) {
+                        Expression expr = (Expression) astNode;
+                        expressionReduce(plainSelect, expr, (select, expression) -> {
+                            try {
+                                nodeSetter.invoke(select, expression);
+                                updateStatements();
+                            } catch (IllegalAccessException | InvocationTargetException e) {
+                                e.printStackTrace();
+                            }
+                        });
+                    } else if (astNode instanceof List) {
+                        List<?> elms = (List<?>) astNode;
+                        if (elms.size() <= 1) {
```

Comment:
Can we sometimes reduce lists with one element to lists with zero elements? In some DBMSs, this is valid. See https://www.db-fiddle.com/f/vXcJLN119VekRuz1jJuyJ5/0

Reply:
OK, I think the 1-element constraint could be removed. :)